unit_tests/kernels/UnitTestFaceElemBasic.C

```diff
@@ -0,0 +1,229 @@
+    for(const stk::mesh::Bucket* bptr : nodeBuckets) {
+        for(stk::mesh::Entity node : *bptr) {
+            double* idptr = stk::mesh::field_data(*idField, node);
+            *idptr = bulk.identifier(node);
+        }
+    }
+}
+
+void verify_faces_exist(const stk::mesh::BulkData& bulk)
+{
+    EXPECT_TRUE(bulk.buckets(stk::topology::FACE_RANK).size() > 0);
+}
+
+class TestFaceElemKernel {
+public:
+    TestFaceElemKernel(stk::topology faceTopo, stk::topology elemTopo,
+                       IdFieldType* idField,
+                       sierra::nalu::ElemDataRequests& faceDataNeeded,
+                       sierra::nalu::ElemDataRequests& elemDataNeeded)
+    : numTimesExecuted_(0), faceTopo_(faceTopo), elemTopo_(elemTopo), idField_(idField)
+    {
+        faceDataNeeded.add_gathered_nodal_field(*idField, 1);
+        elemDataNeeded.add_gathered_nodal_field(*idField, 1);
+    }
+
+    void execute(/*perhaps bundle args into something like SharedMemData... */
+                 sierra::nalu::ScratchViews<DoubleType>& faceViews,
+                 sierra::nalu::ScratchViews<DoubleType>& elemViews,
+                 int numSimdFaces,
+                 const stk::mesh::ConnectivityOrdinal* elemFaceOrdinals)
+    {
+        sierra::nalu::SharedMemView<DoubleType*>& faceNodeIds = faceViews.get_scratch_view_1D(*idField_);
+        sierra::nalu::SharedMemView<DoubleType*>& elemNodeIds = elemViews.get_scratch_view_1D(*idField_);
+        EXPECT_EQ(faceTopo_.num_nodes(), faceNodeIds.size());
+        EXPECT_EQ(elemTopo_.num_nodes(), elemNodeIds.size());
+
+        std::vector<int> faceNodeOrdinals(faceTopo_.num_nodes());
+
+        for(int simdIndex=0; simdIndex<numSimdFaces; ++simdIndex) {
+           elemTopo_.face_node_ordinals(elemFaceOrdinals[simdIndex], faceNodeOrdinals.begin());
+           for(unsigned i=0; i<faceTopo_.num_nodes(); ++i) {
+               DoubleType faceNodeId = faceNodeIds(i);
+               DoubleType elemNodeId = elemNodeIds(faceNodeOrdinals[i]);
+               EXPECT_NEAR(stk::simd::get_data(faceNodeId,simdIndex), stk::simd::get_data(elemNodeId,simdIndex), 1.e-9);
+           }
+        }
+
+        ++numTimesExecuted_;
+    }
+
+    unsigned numTimesExecuted_;
+private:
+    stk::topology faceTopo_;
+    stk::topology elemTopo_;
+    IdFieldType* idField_;
+};
+
+constexpr int simdLen = stk::simd::ndoubles;
+
+int
+calculate_shared_mem_bytes_per_thread(int lhsSize, int rhsSize, int scratchIdsSize, int nDim,
+                                      sierra::nalu::ElemDataRequests& faceDataNeeded,
+                                      sierra::nalu::ElemDataRequests& elemDataNeeded)
+{
+    int bytes_per_thread = (rhsSize + lhsSize)*sizeof(double) + (2*scratchIdsSize)*sizeof(int) +
+                           sierra::nalu::get_num_bytes_pre_req_data<double>(faceDataNeeded, nDim)
+                           +
+                           (rhsSize + lhsSize)*sizeof(double) + (2*scratchIdsSize)*sizeof(int) +
```

Comment:
I just noticed this bytes_per_thread calculation is wrong... I'll add a commit which fixes it.